Base file generation

wtcross opened the pull request:
Base generator for Incubator projects. Sets up for our build pipeline.

Commit:
Base file generation.

Running `yo incubator` and then answering prompts will generate:
- project files like package.json, Gruntfile.js
- dotfiles like .travis.yml, .gitignore, .jshintrc, etc

Commit:
Fixed lint and style

jagoda commented on .editorconfig:
```diff
@@ -0,0 +1,13 @@
+# http://editorconfig.org
```
What is this file for? Is this really something that should be shared with other devs?

wtcross replied:
It's nice to have but now that you mention this I think it conflicts with the philosophy of what we are trying to do with this project. Will remove it, but it might be good for devs to know about http://editorconfig.org/

jagoda replied:
Ah, that's a cool idea -- I had not seen that before.

jagoda commented on .gitattributes:
```diff
@@ -0,0 +1 @@
+* text=auto
```
What does this do?

wtcross replied:
This is generated by the Yeoman generator-generator project. I'll find out what it does.

wtcross replied:
Apparently this ensures normalized LF usage in all files.
https://www.kernel.org/pub/software/scm/git/docs/gitattributes.html

jagoda replied:
Oh, very nice.

jagoda commented on .jshint.json:
```diff
@@ -0,0 +1,21 @@
+  "latedef": true,
+  "newcap": true,
+  "noarg": true,
+  "quotmark": "single",
```
This is in conflict with the JSCS rules. I think that this should be "double".

jagoda commented on .jshint.json:
```diff
@@ -0,0 +1,21 @@
+  "newcap": true,
+  "noarg": true,
+  "quotmark": "single",
+  "regexp": true,
```
What does this do?

wtcross replied:
Rather than placing config for JSCS and JSHint I thought it was a good idea to put things that will never change in a separate file, rather than bloating Gruntfile.

JSHint options set in Gruntfile will be merged with those in .jshint.json, but options set for JSCS will not be. Basically I didn't like how much JSON was in Gruntfile when these tools typically use rc file config

jagoda replied:
Sorry, meant the `regexp` option -- didn't see it when I tried to look for it in the docs.

jagoda commented on .jshint.json:
```diff
@@ -0,0 +1,21 @@
+  "strict": true,
+  "trailing": true,
+  "smarttabs": true,
+  "white": true
```
These rules are not the same as the ones generated by the template. It seems to me that since this is an incubation project the generated rules and the rules used here should be the same unless something prohibits it. For example, these rules are missing useful things like `nonew`.

wtcross replied:
Agreed. Will fix it

jagoda commented on .travis.yml:
```diff
@@ -0,0 +1,8 @@
+  - "0.10"
+before_install:
+  - currentfolder=${PWD##*/}
+  - if [ "$currentfolder" != "generator-incubator" ]; then cd .. && eval "mv $currentfolder generator-incubator" && cd generator-incubator; fi
```
Why is this needed? Shouldn't the current folder just be whatever the project name is? If there is a problem here, I don't think that we really want the build to suppress it, but maybe I'm missing something.

jagoda commented on Gruntfile.js:
```diff
@@ -0,0 +1,47 @@
+	var allFiles    = sourceFiles.concat(testFiles);
+
+	grunt.initConfig({
+		pkg : grunt.file.readJSON("package.json"),
```
What is this used for?

wtcross replied:
It's not specifically used for anything at this time.

jagoda replied:
Can we remove it if we don't need it?

jagoda replied:
Any reason this can't or shouldn't be removed?

wtcross replied:
Not really. Will remove.

jagoda commented on Gruntfile.js:
```diff
@@ -0,0 +1,47 @@
+			}
+		},
+
+		clean : [ "test/temp" ]
```
Why are there temp files? Is this a yeoman thing?

wtcross replied:
The temp directory is created by running tests. I think it might be a good idea to use /tmp instead. This will alleviate the need for grunt-contrib-clean

jagoda replied:
Sounds good if it's easy to do. I generally prefer to have this kind of clean-up in the test suites themselves too so that we have a practice that attempts to avoid undetected test interactions.

wtcross replied:
Fixed.

Commit:
Use bandwidth labs private NPM repo for node builds

jagoda commented on app/index.js:
```diff
@@ -0,0 +1,93 @@
+			name    : "projectName",
+			message : "What is the name of this project?",
+			default : function () {
+				return process.cwd().split(path.sep).pop();
```
I think `path.basename()` will do some of this for you. See http://nodejs.org/api/path.html#path_path_basename_p_ext

jagoda commented on app/index.js:
```diff
@@ -0,0 +1,93 @@
+		this.mkdir("test");
+		this.mkdir("coverage");
+
+		this.copy("editorconfig", ".editorconfig");
```
Is this really something that everybody will want?

jagoda commented on app/templates/_Gruntfile.js:
```diff
@@ -0,0 +1,65 @@
+		},
+
+		jshint : {
+			jshintrc : ".jshint.json",
```
We will want a way to have different lint rules for production code and test code. This is mainly due to frameworks injecting globals fro testing, but there are other minor issues as well.

wtcross replied:
Do you think it will be a good idea to have rules that are used everywhere in these config files? I would prefer to not have everything in Gruntfile so it's easier on the eyes.

jagoda replied:
I think using config files is fine. We just need a way to have different rules for test and production. There are pros and cons to each approach. In my experience, config files leads to duplication where as grunt based lint config allows for inheritance of rules (but leads to a huge gruntfile). If JSHint is allowed to auto-detect the config file I believe you can provide customizations by placing differenc `.jshintrc` files in different directories. See http://www.jshint.com/docs/.

wtcross replied:
That sounds ideal to me as well. This is the only thing left (I think) for this PR. Will work on that tomorrow after talking with you about managing this kind of stuff.

jagoda commented on app/index.js:
```diff
@@ -0,0 +1,93 @@
+	// Where conflicts are handled (used internally)
+	conflicts : function () {
+
+	},
```
Is it possible to just leave these out since they are empty or will things break if we do that?

wtcross replied:
I left them in to give other devs that look at it an idea of what is going on. I can leave them out if you think this is bad style.

jagoda replied:
I'm generally not in favor of leaving code around for the sake of examples. I find this leads to copying code without actually understanding it and it's more to read and keep up with. All that to say, my vote would be that we take it out but open to discussing further.

jagoda commented on app/templates/_package.json:
```diff
@@ -0,0 +1,28 @@
+	"scripts": {
+		"test": "grunt"
+	},
+  "dependencies": {
```
I don't like the idea of having a template bound to specific versions of dependencies. What do you think about scripting calls to `npm install` instead? This way new projects are created with the latest versions of each dependency.

wtcross replied:
I was on the fence on this one. I see value in knowing that everything works with the versions that we are up to date with, but that requires bumping versions in the generator. I'll make this change and see how we like it.

jagoda replied:
Fair enough.

jagoda commented on app/templates/jshint.json:
```diff
@@ -0,0 +1,29 @@
+  "regexp": true,
+  "undef": true,
+  "unused": true,
+  "strict": false,
```
I'm inclined to say that we should force strict mode with `"strict" : true` and `"globalstrict" : true`

jagoda commented on app/templates/travis.yml:
```diff
@@ -0,0 +1,15 @@
+before_install:
+  - npm config set registry https://npm.bwrnd.com
+  - npm config set always-auth true
+  - echo -ne "${NPM_USER}\n${NPM_PASS}\n${NPM_EMAIL}\n" | npm login
```
Clever :-)

wtcross replied:
Apparently not :-(

jagoda commented on app/templates/_package.json:
```diff
@@ -0,0 +1,28 @@
+    "grunt-jscs-checker": "^0.4.4",
+    "grunt-mocha-test": "^0.11.0",
+    "grunt-istanbul-coverage": "0.0.5"
+	}
```
I think that we should add `"private" : true` to this template. See https://www.npmjs.org/doc/json.html#private

jagoda commented on app/templates/editorconfig:
```diff
@@ -0,0 +1,13 @@
+# http://editorconfig.org
```
Is this really a file that should be shared?

jagoda commented on test/creation.js:
```diff
@@ -0,0 +1,43 @@
+/*global describe, beforeEach, it */
```
I think these should be in a jshint config that applies to all tests rather than replicated across every test file.

wtcross replied:
Agreed.

jagoda commented on test/creation.js:
```diff
@@ -0,0 +1,43 @@
+			done();
+		}.bind(this));
+	});
+
```
I think it is generally a better practice to have tests clean-up after themselves using an `after()` or `afterEach()` clause rather than relying on the orchestration tool for this. This reduces the likelihood of tests interacting with each other.

wtcross replied:
Fixed.

jagoda commented on test/loading.js:
```diff
@@ -0,0 +1,9 @@
+/*global describe, beforeEach, it*/
```
I think these should be in a jshint config that applies to all tests rather than replicated across every test file.

jagoda commented on app/templates/_package.json:
```diff
@@ -0,0 +1,28 @@
+    "q": "^1.0.1"
+	},
+	"devDependencies": {
+    "mocha": "^1.20.0",
```
I think that we should provide and specify an assertion library. I prefer [chai](http://chaijs.com/) but I'm open to discussing other options.

jagoda commented on test/loading.js:
```diff
@@ -0,0 +1,9 @@
+/*global describe, beforeEach, it*/
+var assert = require("assert");
+
+describe("incubator generator", function () {
```
Since this `describe` has the same description as the one in `creation.js`, do you think it would create a better developer experience to merge these test suites? I'm thinking that this could be very tedious when debugging failures since the same description can appear across multiple files.

wtcross replied:
Agreed. I made this change.

jagoda commented on test/creation.js:
```diff
@@ -0,0 +1,43 @@
+				"../../app"
+			]);
+			done();
+		}.bind(this));
```
Can you explain what the `bind(this)` does for me? This looks like it is binding a function to the mocha test runner. Is this intentional?

wtcross replied:
This is necessary to use Yeoman's helpers from what I can see.

jagoda replied:
Is it a requirement of the API or because you have `this.app = . . .` in line 12? In either case, binding to the mocha test runner and assigning properties to it is probably not ideal -- I would expect us to at least bind to something that either we created or yeoman expects.

wtcross replied:
No longer an issue- I fixed it. I made myself into a victim of copy + paste!

Commit:
Changes for PR 1

wtcross commented:
I fixed everything except for test directory cleanup. I'll add an afterEach that will clean up and use /tmp instead of a temp directory in the repo. Also, I'll add global JSHint test rules.

wtcross commented:
I broke the build! I forgot to add Q to the generator dependencies.

Commit:
Added Q to package.json

jagoda commented on .travis.yml:
```diff
@@ -0,0 +1,17 @@
+  - npm config set always-auth true
+  - echo -ne "${NPM_USER}\n${NPM_PASS}\n${NPM_EMAIL}\n" | npm login
+  - currentfolder=${PWD##*/}
+  - if [ "$currentfolder" != "generator-incubator" ]; then cd .. && eval "mv $currentfolder generator-incubator" && cd generator-incubator; fi
```
What is this for? I think I had a comment on this in a previous change that was not addressed.

wtcross replied:
It was put there by generator-generator when I created this project. I removed it since it looks unnecessary.

jagoda commented:
Are you intending to add more commits? It looked liked several of my comments still have not been addressed. I'd like to at least discuss everything before merging.

wtcross commented:
Definitely going to add more commits. Wanted to get what I had our there
while I work on the rest.
On Jun 4, 2014 1:27 PM, "Jeffrey Jagoda" notifications@github.com wrote:

> Are you intending to add more commits? It looked liked several of my
> comments still have not been addressed. I'd like to at least discuss
> everything before merging.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/inetCatapult/generator-incubator/pull/1#issuecomment-45122860
> .

jagoda commented:
Ah, cool -- makes total sense. Sorry for the confusion.

Commit:
No longer bind to mocha scope

Commit:
Add rimraf to package.json dependencies

wtcross commented:
#### TODO:
- [x] JSHint config organization
- [x] JSCS config organization
- [x] Mocha JSHint rules

I'd like to have a conversation about config organization for JSHint and JSCS.

Commit:
JSHint configuration management

- renamed .jscs.json to .jscsrc to follow conventions
- cleaned up root .jshint.json and test .jshint.json
- .jshint.json extends .jshint.json with globals for mocha
- updated tests to check for these files

wtcross commented:
I went ahead and made these changes. JSHint is now running correctly on all files. Let me know if everything looks good.

Commit:
Added lodash to package.json -_-

Commit:
Make use of a new istanbul plugin for grunt.

jagoda commented:
Looks good overall -- had a couple of minor questions. Is the plan still to add coverage enforcement?

Commit:
Refactor of generator design.

- no longer use "extend" design pattern.
- main generator will only print greeting message and invoke sub-generators
- added "core" generator (future generators might be web,cli,travis,etc.)

wtcross commented:
Forgot to push changes I made last week. Let me know what you think. Here is a summary of what I did:
- no longer use extend design pattern
- main "app" generator just prints a welcome message and invokes required sub-generators
- added an opt-in watch task for grunt

Commit:
Updated readme, fixed bug with git repo url

Commit:
Fixed bug in core test case

Commit:
Fixes recommended by Jeff. Disabled coverage checking for this project.

Commit:
Use JSON.parse rather than require when loading JSHint options

jagoda commented:
This looks good to me. I'm fine with merging once Travis is happy with these changes.

Commit:
Require fs module in Gruntfile
